GOALS.md

```diff
@@ -0,0 +1,233 @@
+  [KxStudio](https://kx.studio) project has a few [extra specifications](https://kx.studio/ns/) that some plugins
+  implement. The [Ardour](https://ardour.org) DAW also has some non-standard specifications for their inline strip
+  displays, which can be quite useful.
+
+## Lightweight, fast library
+
+Obviously, any library handling realtime audio needs to be fast to stay within the time budgets of the given audio
+buffer sizes (and avoid XRuns).
+
+This is where Rust's "zero-cost abstractions" truly shine, as we can build higher-level abstractions that produce little
+to no extra machine code to execute (and thus a minimal performance penalty). Of course, the `lv2` crate provides a
+large majority of those in its APIs, a vast amount being nothing but a wrapper around the pointers given to the plugin
+by the host.
+
+However, there is another important performance consideration that is not just "how fast can we do the processing":
+plugin implementations need to be as lightweight as possible. This means than both plugin authors, and the libraries
+they use (such as the `lv2`), must be *extremely* conservative about the resources they allocate for themselves.
+Whether it is memory, threads, or other synchronization primitives that can introduce delays or locking.
+
+The reason for this is simple: a plugin is very rarely alone when it's used in a DAW. It is most likely to run alongside
+dozens, if not hundreds, of other plugins. Not to mention the processing the DAW itself needs to do for mixing all of
+these, and the I/O it needs to perform to the hardware to get actual sound. All of it in very tight timing budgets.
+
+For instance, audio processing code may want to spread work across multiple threads to be as fast as possible. This
+works fine in a standalone application where the process can allocate most or all of the CPU cores to itself. However,
+this can't work in a session where there are many instances of that code competing for CPU power at the same time. The
+OS would have to interrupt and reschedule threads constantly, losing most of the CPU power in context switches.
+
+In general, the LV2 APIs consider the host to be in charge of handling most of the plugin's resources, and behave
+like a scheduler or executor of sorts for the various plugins. Because it knows the state of all the plugins running,
+as well as their (potentially complex) I/O configuration, it can apply massive optimizations to its scheduling and
+processing, such as spreading the work on a single thread pool. However, this means the plugins must cooperate, and
+cannot do what they want. Examples of work LV2 plugins should defer to the host are I/O buffers and communication, state
+serialization (i.e., presets / session loading and saving), asynchronous processing, or UI communication.
+
+In that aspect, LV2 plugins and hosts share very similar behaviors with Rust's own
```

Comment:
I'm not familiar with futures and executors (i didn't ), so I don't fully understand this section. Would it be possible to reword this section so that people can understand it without knowledge about futures and executors?

Reply:
Yes, re-reading this I realize this is a bit more of me rambling about LV2's API (compared to others like VST) more than anything else. I will reword (if not completely remove), as it's more of an example to illustrate the rest of this section.